### PR TITLE
Cu 2y216vn validationrs asset does not have an associated vault validity check refactor mike belous

### PR DESCRIPTION
**Pull request by belousm**
## Issue
- https://app.clickup.com/t/2y216vn
- https://app.clickup.com/t/2y216xy
## Description
add_liquidity and remove_liquidity refactoring

## Checklist

- [x] I have updated the cargo docs to reflect changes made by this PR _(if applicable)_
- [x] I have updated the `book/` to reflect changes made by this PR _(if applicable)_

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2y216vn Validation.rs AssetDoesNotHaveAnAssociatedVault validity check refactor](https://app.clickup.com/t/2y216vn)

File: frame/instrumental/src/lib.rs [hunk 1]
```diff
@@ -75,7 +75,7 @@ pub mod pallet {
 	// -------------------------------------------------------------------------------------------
 
 	use crate::weights::WeightInfo;
-	use crate::validation::ValidateVaultDoesNotExist;
+	use crate::validation::{ValidateVaultDoesNotExist, ValidateVaultDoesExists};
```
Comment (kevin-composable) [inline frame/instrumental/src/lib.rs:78]: can you rename to ValidateVaultDoesExist

**Review by kevin-composable — APPROVED** [review]

File: frame/instrumental/src/lib.rs [hunk 2]
```diff
@@ -446,14 +446,10 @@ pub mod pallet {
...
 			// Requirement 1) The asset must have an associated vault
-			ensure!(
-				AssetVault::<T>::contains_key(asset),
-				Error::<T>::AssetDoesNotHaveAnAssociatedVault
-			);
-
-			Self::do_add_liquidity(issuer, asset, amount)?;
-
-			Ok(())
+			match Validated::<&T::AssetId, ValidateVaultDoesExists<T>>::new(asset) {
```
Comment (saruman9) [inline frame/instrumental/src/lib.rs:449]: Just `Validated::new` (see above).

File: frame/instrumental/src/lib.rs [hunk 3]
```diff
@@ -478,14 +474,10 @@ pub mod pallet {
...
 			// Requirement 1) The asset must have an associated vault
-			ensure!(
-				AssetVault::<T>::contains_key(asset),
-				Error::<T>::AssetDoesNotHaveAnAssociatedVault
-			);
-
-			Self::do_remove_liquidity(issuer, asset, amount)?;
-
-			Ok(())
+			match Validated::<&T::AssetId, ValidateVaultDoesExists<T>>::new(asset) {
```
Comment (saruman9) [inline frame/instrumental/src/lib.rs:477]: Just `Validated::new` (see above).

File: frame/instrumental/src/lib.rs [hunk 4]
```diff
@@ -418,7 +418,7 @@ pub mod pallet {
 		fn create(
 			config: InstrumentalVaultConfigFor<T>
 		) -> Result<Self::VaultId, DispatchError> {
-			match Validated::new(config) {
+			match Validated::<InstrumentalVaultConfigFor<T>, ValidateVaultDoesNotExist<T>>::new(config) {
```
[the remark this answers is not in the document]
Comment (belousm) [reply frame/instrumental/src/lib.rs:421]: I see. Change it.

**Pull request merged**